pandas/core/reshape/merge.py

```diff
@@ -2219,6 +2219,13 @@ def _factorize_keys(
         # "_values_for_factorize"
         rk, _ = rk._values_for_factorize()  # type: ignore[union-attr]
 
+    elif is_integer_dtype(lk.dtype) and is_integer_dtype(rk.dtype):
```

Comment:
following yesterday's call we are in a position to do this The Right Way.

what you're doing here in the general case iff rk.dtype == lk.dtype.  otherwise you need to cast to do find_common_type first.

i think this is OK for our existing IntegerArray, but it is definitely relying on implementation details.

Reply:
Could you clarify what the right way means?

Reply:
The documented requirements for values_for_factorize (vff) don't say anything (yet) about how left.vff() relates to right.vff(). Without pinning those down, we can cook up cases where what you're doing here would produce nonsense results.  As long as you're restricting to our internal IntegerArray, this should be fine.